### DISCUSSION TITLE
High Throughput MQTT

### DISCUSSION THREAD

**Discussion by mohdali** [elsa-workflows/elsa-core#3057] [Design & tech]
Hi,

I would like to check the capability for Elsa and MQTT activities to handle high throughput data.

Let's say I have multiple MQTT producers that are going to send multiple data points readings every 5 seconds. Each producer will send 50+ readings. I would like then to create a workflow that collects those readings and insert them into a database. Ideally using BULK inserts for higher performance.

I understand the current MQTT activity will trigger for each message received, which will result in lots of instances being created and persisted in history. Is there an alternative way to implement this to avoid generating a new instance for each message? Probably work like a stream processor?

I have implemented this earlier using Azure IoT Hub but would really like to replicate it in Elsa.

**Comment by sfmskywalker** [comment]
One way could be to make sure your workflow instance is correlated to some value, and have your message consumer trigger this particular workflow instance for every message that comes in. Your workflow should make sure that after processing a message, it connects back to your MQTT activity.

You could do it without correlation too, but it would require you to manually manage a workflow instance (create it and then resume it every time a message comes in).

One disadvantage of doing this with Elsa 2 is the fact that it needs to write the workflow instance back to the database - unless of course you use the memory provider or "null" provider where nothing gets persisted.

Elsa 3 will have a new type of activity called **generator** - which is what you might want for your scenario.
A generator is an activity that is similar to a trigger, but instead of creating new workflow instances, it executes the workflow instance it is a part of. 

For example, the `Timer` activity will enable scenarios where you have a workflow with a timer somewhere on your workflow, which "ticks" at a regular interval, regardless of where your workflow currently is. 

With regards to handling high throughput data, Elsa 3 can further enhance this by having better control over how often a workflow instance is updated in the database. For example, instead of saving the workflow instance to the database every time a message comes in, this could be done once every 5 minutes, significantly increasing throughput.

This is made possible with ProtoActor, where each workflow instance is modeled as an actor that remains in memory as messages come in. This way, we could even offer hosting runtimes that never update workflow instances, but simply process messages from e.g. Kafka, and perhaps take a snapshot once in a while. Or perhaps workflow state needs to be "exported" to e.g. some indexer like Elastic Search in some format that makes it easy to look for workflows.

**Reply by mohdali** [reply]
Hi @sfmskywalker, thanks for the detailed response. I'm planning to create a prototype for this in the near future. Would be great to evaluate how to do that in both Elsa 2 and Elsa 3 as well :)

> One way could be to make sure your workflow instance is correlated to some value, and have your message consumer trigger this particular workflow instance for every message that comes in. Your workflow should make sure that after processing a message, it connects back to your MQTT activity.

Is this possible with the current MQTT activity or does it require creating another custom activity for MQTT? probably the correlation Id can become the device name and then we have one instance per device. But can I set that before the instances are created? So probably every time a new device is added, new instance is created?

Also how to do this kind of looping back? I couldn't define a loop in the designer. Is it possible in Elsa 2?

> You could do it without correlation too, but it would require you to manually manage a workflow instance (create it and then resume it every time a message comes in).

Any example available for this? I haven't tried it before.

> One disadvantage of doing this with Elsa 2 is the fact that it needs to write the workflow instance back to the database - unless of course you use the memory provider or "null" provider where nothing gets persisted.

Can this be set in Elsa 2 at a workflow or workflow provider level? Otherwise, it seems that maybe I need a dedicated Elsa instance that has memory or null provider to avoid persisting an instance per message.

> Elsa 3 will have a new type of activity called **generator** - which is what you might want for your scenario. A generator is an activity that is similar to a trigger, but instead of creating new workflow instances, it executes the workflow instance it is a part of.
> 
> For example, the `Timer` activity will enable scenarios where you have a workflow with a timer somewhere on your workflow, which "ticks" at a regular interval, regardless of where your workflow currently is.
> 
> With regards to handling high throughput data, Elsa 3 can further enhance this by having better control over how often a workflow instance is updated in the database. For example, instead of saving the workflow instance to the database every time a message comes in, this could be done once every 5 minutes, significantly increasing throughput.
> 
> This is made possible with ProtoActor, where each workflow instance is modeled as an actor that remains in memory as messages come in. This way, we could even offer hosting runtimes that never update workflow instances, but simply process messages from e.g. Kafka, and perhaps take a snapshot once in a while. Or perhaps workflow state needs to be "exported" to e.g. some indexer like Elastic Search in some format that makes it easy to look for workflows.

Awesome, looks Like Elsa 3 has more flexibility to handle this scenario better. I would definitely want to see the way to transition to Elsa 3 while evolving this prototype.
